loopy/transform/loop_fusion.py

```
@@ -0,0 +1,576 @@
+    """
+    from loopy.kernel.data import ConcurrentTag
+    from loopy.schedule.tools import (
+        _get_partial_loop_nest_tree,
+        _get_iname_to_tree_node_id_from_partial_loop_nest_tree)
+
+    assert isinstance(kernel, LoopKernel)
+    assert isinstance(candidates, frozenset)
+
+    vng = kernel.get_var_name_generator()
+    fused_chunks = {}
+
+    # {{{ handle concurrent inames
+
+    # filter out concurrent loops.
+    all_concurrent_tags = reduce(frozenset.union,
+                                 (kernel.inames[iname].tags_of_type(ConcurrentTag)
+                                  for iname in candidates),
+                                 frozenset())
+
+    concurrent_tag_to_inames = {tag: set()
+                                for tag in all_concurrent_tags}
+
+    for iname in candidates:
+        if kernel.inames[iname].tags_of_type(ConcurrentTag):
+            # since ConcurrentTag is a UniqueTag there must be exactly one of
+            # it.
+            tag, = kernel.tags_of_type(ConcurrentTag)
+            concurrent_tag_to_inames[tag].add(iname)
+
+    for inames in concurrent_tag_to_inames.values():
+        fused_chunks[vng(prefix)] = inames
+        candidates = candidates - inames
+
+    # }}}
+
+    tree = _get_partial_loop_nest_tree(kernel)
+    iname_to_tree_node_id = (
+        _get_iname_to_tree_node_id_from_partial_loop_nest_tree(tree))
+
+    # {{{ sanitary checks
+
+    _nest_tree_id_to_candidate = {}
+
+    for iname in candidates:
+        loop_nest_tree_node_id = iname_to_tree_node_id[iname]
+        if loop_nest_tree_node_id not in _nest_tree_id_to_candidate:
+            _nest_tree_id_to_candidate[loop_nest_tree_node_id] = iname
+        else:
+            conflict_iname = _nest_tree_id_to_candidate[loop_nest_tree_node_id]
+            raise LoopyError(f"'{iname}' and '{conflict_iname}' "
+                             "cannot fused be fused as they can be nested "
+                             "within one another.")
```

Does this need to return an error or could we just remove `iname` from `candidates` in this case?

---

The idea was that this needs to be an error and the caller passes in the candidates appropriately so that the loop nests can be fused with each other.

---

Got rid of it.